KAFKA-9555 Added default RLMM implementation based on internal topic storage.

satishd opened the pull request:
KAFKA-9555 Added default RLMM implementation based on internal topic storage.
- This is the initial version of the default RLMM implementation. 
- This includes changes containing default RLMM configs, RLMM implementation, producer/consumer managers.
- Introduced `TopicBasedRemoteLogMetadataManagerHarness` which takes care of bringing up a Kafka cluster and create remote log metadata topic and initializes `TopicBasedRemoteLogMetadataManager`.
- Refactored existing `RemoteLogMetadataCacheTest` to `RemoteLogSegmentLifecycleTest` to have parameterized tests to run both `RemoteLogMetadataCache` and also `TopicBasedRemoteLogMetadataManager`.
- Refactored existing `InmemoryRemoteLogMetadataManagerTest`,  `RemoteLogMetadataManagerTest` to have parameterized tests to run both `InmemoryRemoteLogMetadataManager` and also `TopicBasedRemoteLogMetadataManager`.

We will have a followup PR adding file based cache for storing the remote log state for each partition to avoid consuming from the remote log metadata topic from the earliest whenever a broker is restarted.

This is part of tiered storage KIP-405 efforts.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

satishd committed:
KAFKA-9555 Added default RLMM implementation based on internal topic storage.
Introduced TopicBasedRemoteLogMetadataManagerHarness which takes care of bringing up a Kafka cluster and create remote log metadata topic and initializes TopicBasedRemoteLogMetadataManager.
Added TopicBasedRemoteLogMetadataManager to run the existing tests by making the existing RemoteLogMetadataManagerTest parameterized.

This is part of tiered storage KIP-405 efforts.

satishd commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManagerConfig.java:
```diff
@@ -0,0 +1,197 @@
+
+    private static final ConfigDef CONFIG = new ConfigDef();
+    static {
+        CONFIG.define(REMOTE_LOG_METADATA_TOPIC_REPLICATION_FACTOR_PROP, INT, DEFAULT_REMOTE_LOG_METADATA_TOPIC_REPLICATION_FACTOR, atLeast(1), LOW,
```
`Config` definition will be moved to `KafkaConfig` later when default RLMM is integrated with the broker.

satishd commented:
@junrao @kowshik This PR is ready for review. Pl let me know your comments.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManager.java:
```diff
@@ -0,0 +1,346 @@
+
+    private void initializeResources() {
+        log.info("Initializing the resources.");
+        while (true) {
```
If the producer/consumer are configured incorrectly, we want to fail faster instead of retrying.

satishd replied:
That makes sense to me.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManager.java:
```diff
@@ -0,0 +1,346 @@
+                    // We may want to add a check here later.
+
+                    // Create producer and consumer managers.
+                    if (producerManager == null) {
```
Since producerManager and consumerManager are updated in a separate thread without holding lock, do they need to be volatile?

satishd replied:
Making volatile makes sense to me.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManager.java:
```diff
@@ -0,0 +1,346 @@
+    }
+
+    // Visible for testing.
+    void setTime(Time time) {
```
Should we pass in time through the constructor?

satishd replied:
Done.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManager.java:
```diff
@@ -0,0 +1,346 @@
+    private void doPublishMetadata(TopicIdPartition topicIdPartition, RemoteLogMetadata remoteLogMetadata)
+            throws RemoteStorageException {
+        log.debug("Publishing metadata for partition: [{}] with context: [{}]", topicIdPartition, remoteLogMetadata);
+        ensureInitializedAndNotClosed();
```
Do we need this since it's in the caller already?

satishd replied:
Removed the check as you suggsted.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManager.java:
```diff
@@ -0,0 +1,346 @@
+                                                                           remoteLogMetadata);
+            // Wait until the consumer catches up with this offset. This will ensure read-after-write consistency
+            // semantics.
+            consumerManager.waitTillConsumptionCatchesUp(recordMetadata);
```
Waiting for each event to be consumed reduces throughput. Could we handle the expected metadata load with this?

satishd replied:
This method is invoked from multiple threads for different topic partitions. This will not be a bottleneck as each partition's segments will be uploaded in a sequential manner.

kowshik replied:
I agree with the question here. This can become more expensive than it seems. The alternative is to pursue an asynchronous notification model to improve the throughput.

satishd replied:
As we discussed offlime, this may not become a bottleneck but we will make respective RLMM APIs asynchronous so that the APIs are extensible and implementors can provide async behavior.FIled [KAFKA-12988](https://issues.apache.org/jira/browse/KAFKA-12988) to track this issue.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ProducerManager.java:
```diff
@@ -0,0 +1,125 @@
+    }
+
+    public RecordMetadata publishMessage(TopicIdPartition topicIdPartition,
+                                         RemoteLogMetadata remoteLogMetadataContext) throws KafkaException {
```
remoteLogMetadataContext => remoteLogMetadata?

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ProducerManager.java:
```diff
@@ -0,0 +1,125 @@
+        if (callback.exception() != null) {
+            Exception ex = callback.exception();
+            if (ex instanceof KafkaException) {
+                throw (KafkaException) ex;
```
No need to cast to KafkaException.

satishd replied:
It is needed as the method is declared with throws KafkaException. I am also fine with removing it as it is a RTE.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ProducerManager.java:
```diff
@@ -0,0 +1,125 @@
+                                         " is more than the partition count: " + rlmmConfig.metadataTopicPartitionsCount());
+            }
+            producer.send(new ProducerRecord<>(rlmmConfig.remoteLogMetadataTopicName(), metadataPartitionNo, null,
+                    serde.serialize(remoteLogMetadataContext)), callback).get();
```
Sending one event at a time reduces the batching benefit in the producer. Could this handle the expected metadata load?

satishd replied:
Multiple events are published by multiple threads and batching will occur in the producer.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java:
```diff
@@ -0,0 +1,136 @@
+import java.util.Optional;
+import java.util.Set;
+
+public class ConsumerManager implements Closeable {
```
Could we add a comment for this class?

satishd replied:
I will add that.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -0,0 +1,253 @@
+ * partitions to be deleted. This class receives those notifications with
+ * {@link #addAssignmentsForPartitions(Set)} and {@link #removeAssignmentsForPartitions(Set)} assigns consumer for the
+ * respective remote log metadata partitions by using {@link RemoteLogMetadataTopicPartitioner#metadataPartition(TopicIdPartition)}.
+ * Any leadership changes later are called through the same API. We will remove the partitions that ard deleted from
```
typo ard

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -0,0 +1,253 @@
+            Set<TopicIdPartition> updatedReassignedPartitions = new HashSet<>(assignedTopicPartitions);
+            updatedReassignedPartitions.addAll(addedPartitions);
+            updatedReassignedPartitions.removeAll(removedPartitions);
+            Set<Integer> updatedAssignedMetaPartitions = new HashSet<>();
```
Hmm, shouldn't we start with assignedMetaPartitions?

satishd replied:
Are you saying it should be initialized with `updatedAssignedMetaPartitions = new HashSet<>(assignedMetaPartitions);`. This is not needed as we are recomputing that set based on updatedReassignedPartitions.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -0,0 +1,253 @@
+                updatedAssignedMetaPartitions.add(topicPartitioner.metadataPartition(tp));
+            }
+
+            if (!updatedAssignedMetaPartitions.equals(assignedMetaPartitions)) {
```
Hmm, assignedMetaPartitions may not change if assignedTopicPartitions changes. Should we still update assignedTopicPartitions in that case?

satishd replied:
Nice catch, updated.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -0,0 +1,253 @@
+        Map<TopicPartition, Long> endOffsets;
+        while (true) {
+            try {
+                endOffsets = consumer.endOffsets(assignedMetaTopicPartitions, Duration.ofSeconds(30));
```
From which offset does the consumer start fetching?

satishd replied:
We have not yet added the code to store the consumed offset and start from those offsets whenever the consumer starts fetching from those partitions. We plan to add that in a subsequent PR.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -0,0 +1,253 @@
+                ConsumerRecords<byte[], byte[]> consumerRecords
+                        = consumer.poll(Duration.ofSeconds(POLL_INTERVAL_MS));
+                for (ConsumerRecord<byte[], byte[]> record : consumerRecords) {
+                    handleRemoteLogMetadata(serde.deserialize(record.value()));
```
Should we only process events corresponding to assignPartitions?

satishd replied:
Right, I will add that check.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -0,0 +1,253 @@
+
+            if (!updatedAssignedMetaPartitions.equals(assignedMetaPartitions)) {
+                assignedTopicPartitions = Collections.unmodifiableSet(updatedReassignedPartitions);
+                assignedMetaPartitions = Collections.unmodifiableSet(updatedAssignedMetaPartitions);
```
If a partition is moved to a different broker, we will need to bootstrap the remote state for the partition by consuming from the beginning of the remote metadata partition. How is that handled?

satishd replied:
This is not addressed in this PR. I planned to have a followup PR for these changes. I may use a different consumer for the newly subscribed partitions to build the state.

kowshik replied:
It appears that once this logic is implemented, there is probably no need to wait for `maybeWaitForPartitionsAssignment`. The reason is that whenever a partition is assigned, we will bootstrap the remote state by consuming from the beginning.

ccding replied:
What is the cost of consuming from the beginning if the remote metadata partition grows huge?

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataTopicPartitioner.java:
```diff
@@ -0,0 +1,43 @@
+    public int metadataPartition(TopicIdPartition topicIdPartition) {
+        Objects.requireNonNull(topicIdPartition, "TopicPartition can not be null");
+
+        int partitionNo = Utils.toPositive(Utils.murmur2(topicIdPartition.toString().getBytes(StandardCharsets.UTF_8)))
```
The toString() method could change over time. Perhaps it's safer to compute a customized hashcode for topicIdPartition here.

satishd replied:
Good point. I will add that.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManagerConfig.java:
```diff
@@ -0,0 +1,197 @@
+
+        Map<String, Object> parsedConfigs = CONFIG.parse(props);
+
+        bootstrapServers = (String) props.get(BOOTSTRAP_SERVERS_CONFIG);
```
BOOTSTRAP_SERVERS_CONFIG should be prefixed with REMOTE_LOG_METADATA_CLIENT_PREFIX, right?

satishd replied:
REMOTE_LOG_METADATA_CLIENT_PREFIX is just a prefix for generating client-ids for producer and consumer. If you are talking about the remote log metadata property prefix, It is assumed that the caller would have already removed those prefixes and sent the config. These prefixes are defined [here](https://github.com/apache/kafka/pull/10733/files/f185b4f86b457da35c21f1461fe1b1e1d73e4939#diff-31933b76c0578600db8744507e5c5ff840820d5a181a2a5d1e08c50bdc64ea6fR46)

junrao replied:
Other plugins on the broker may also need a bootstrap_server config. To distinguish them, it would be useful to add a prefix that's specific to remote storage.

junrao replied:
Is this comment addressed?

satishd replied:
`RemoteLogMetadataManager.configure(Map<String, ?> configs)` is always invoked with stripping the RLMM  prefix. "bootstrap.servers" property is sent as part of the configs here and any registered RLMM plugin will receive this property.

junrao replied:
Since bootstrap_servers is used for the internal producer/consumer, should bootstrap_servers be defined with a prefix of  REMOTE_LOG_METADATA_COMMON_CLIENT_PREFIX, REMOTE_LOG_METADATA_COMMON_CLIENT_PREFIX or REMOTE_LOG_METADATA_CONSUMER_PREFIX?

satishd replied:
No, it is not needed to be sent with any prefix(like common_client, producer or consumer) because "bootstrap.servers" property is sent for any registered RLMM plugin but not only limited to the default RLMM. I will update the KIP with these details.

junrao replied:
Hmm, "bootstrap.servers" makes sense for a topic based RLMM since it depends on Kafka. Why do we require "bootstrap.servers" in other RLMM implementations?

satishd replied:
I am not sure about usecases with other RLMM but it allows them to connect to the broker.

junrao replied:
What about client security related properties? It's weird that we pick up "bootstrap.servers" from one prefix, but the corresponding security properties under a different prefix. If we do provide the security related properties in the same prefix, they seem to be duplicated from those under prefix REMOTE_LOG_METADATA_COMMON_CLIENT_PREFIX, REMOTE_LOG_METADATA_COMMON_CLIENT_PREFIX or REMOTE_LOG_METADATA_CONSUMER_PREFIX.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManagerConfig.java:
```diff
@@ -0,0 +1,197 @@
+
+        for (Map.Entry<String, ?> entry : configs.entrySet()) {
+            String key = entry.getKey();
+            if (key.startsWith(REMOTE_LOG_METADATA_COMMON_CLIENT_PREFIX)) {
```
Do we need this? It seems that it's easier to just duplicate the property for producer and consumer.

satishd replied:
This is to avoid duplicate entries for both the producer and consumer. We added that in the KIP earlier. 
If duplicating is the way we use at other places if any, I am fine with that.

kowshik replied:
I had the same question. It appears better to just duplicate the properties.

satishd replied:
As we discussed offline, we see the benefit of keeping several common client config like security to be shared across producer and consumer props avoiding any copy/paste mistakes. 
User has an option not to use common client configs and use the respective producer and consumer configs.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -0,0 +1,253 @@
+                    for (Map.Entry<Integer, Long> entry : partitionToTargetEndOffsets.entrySet()) {
+                        final Long offset = partitionToConsumedOffsets.getOrDefault(entry.getKey(), 0L);
+                        if (offset >= entry.getValue()) {
+                            partitionToTargetEndOffsets.remove(entry.getKey());
```
What's the purpose of waiting for consumption up to partitionToTargetEndOffsets?

satishd replied:
This is leftover code for other changes that I was working on for handling partition moving to a new broker in another branch. I will remove it.

junrao reviewed (commented):
@satishd : Thanks for the PR. Made a pass of non-testing files. A few comments below.

satishd committed:
Addressed review comments

satishd commented:
Thanks @junrao for your comment. Addressed most of them with the latest commit.

satishd committed:
Addressed reveiw comments.
Added transformation of TopicIdPartition into bytes with custom hashing.
Skipping the events which are not currently assigned for the ConsumerTask.

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ProducerManager.java:
```diff
@@ -0,0 +1,125 @@
+
+    private void ensureNotClosed() {
+        if (close) {
+            throw new IllegalStateException("This instance is already set in close state.");
```
s/set in close state/closed

satishd replied:
Done

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ProducerManager.java:
```diff
@@ -0,0 +1,125 @@
+                                         RemoteLogMetadata remoteLogMetadata) throws KafkaException {
+        ensureNotClosed();
+
+        int metadataPartitionNo = topicPartitioner.metadataPartition(topicIdPartition);
```
s/metadataPartitionNo/metadataPartition

satishd replied:
Done

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ProducerManager.java:
```diff
@@ -0,0 +1,125 @@
+    }
+
+    public RecordMetadata publishMessage(TopicIdPartition topicIdPartition,
+                                         RemoteLogMetadata remoteLogMetadata) throws KafkaException {
```
It appears you could eliminate the additional `topicIdPartition` parameter and instead use the value returned by `remoteLogMetadata.topicIdPartition()` API.

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ProducerManager.java:
```diff
@@ -0,0 +1,125 @@
+/**
+ * This class is responsible for publishing messages into the remote log metadata topic partitions.
+ */
+public class ProducerManager implements Closeable {
```
It appears this class does not have unit tests currently. Is there a plan to add unit tests?

satishd replied:
I plan to add that later.

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManagerConfig.java:
```diff
@@ -0,0 +1,197 @@
+import static org.apache.kafka.common.config.ConfigDef.Type.INT;
+import static org.apache.kafka.common.config.ConfigDef.Type.LONG;
+
+public final class TopicBasedRemoteLogMetadataManagerConfig {
```
Could you pls add a comment for this class?

satishd replied:
Done

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManagerConfig.java:
```diff
@@ -0,0 +1,197 @@
+    private final int metadataTopicPartitionsCount;
+    private final String bootstrapServers;
+    private final long consumeWaitMs;
+    private final long metadataTopicRetentionMillis;
```
s/Millis/Ms

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManagerConfig.java:
```diff
@@ -0,0 +1,197 @@
+public final class TopicBasedRemoteLogMetadataManagerConfig {
+    private static final Logger log = LoggerFactory.getLogger(TopicBasedRemoteLogMetadataManagerConfig.class.getName());
+
+    public static final String REMOTE_LOG_METADATA_TOPIC_NAME = "__remote_log_metadata";
```
It seems that we have internal topics specified in `org.apache.kafka.common.internals.Topic` class.
Don't we want this new internal topic to be defined in the `Topic` class, together with other internal topics?

satishd replied:
I plan to add this once RLMM is called from remote log layer classes. I wanted this change to be self contained for now.

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManagerConfig.java:
```diff
@@ -0,0 +1,197 @@
+        props.put(CommonClientConfigs.CLIENT_ID_CONFIG, clientIdPrefix + "_consumer");
+        props.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, false);
+        props.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+        props.put(ConsumerConfig.EXCLUDE_INTERNAL_TOPICS_CONFIG, false);
```
Hmm, why do you need this exclusion to be true?

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManagerConfig.java:
```diff
@@ -0,0 +1,197 @@
+            "Default: -1, that means unlimited. Users can configure this value based on their use cases. " +
+            "To avoid any data loss, this value should be more than the maximum retention period of any topic enabled with " +
+            "tiered storage in the cluster.";
+    public static final String REMOTE_LOG_METADATA_CONSUME_WAIT_MS_DOC = "The amount of time in milli seconds to wait for the local consumer to " +
```
Is this the timeout for how long you'd want the client to wait to consume the message that it produces to `__remote_log_metadata` topic? If yes, then don't we want this timeout to be unlimited i.e. we wait as long as it takes to consume the published event?

satishd replied:
We do not want this to be completely blocked as we want to release the remote log thread after a specific timeout in case of any intermittent issues so that other partitions tiring can proceed.

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManagerConfig.java:
```diff
@@ -0,0 +1,197 @@
+                ", bootstrapServers='" + bootstrapServers + '\'' +
+                ", consumeWaitMs=" + consumeWaitMs +
+                ", metadataTopicRetentionMillis=" + metadataTopicRetentionMillis +
+                ", consumerProps=" + consumerProps +
```
`consumerProps` and `producerProps` are of type `Map`, therefore the `.toString()` is probably not readable. So you'd need to convert these into a comma-separated list sth like `K1=V1,K2=V2,...Kn=Vn`.

satishd replied:
We are using HashMap for these instances and it prints k,v format. Are you suggesting that this map implementation may change as it is of type Map and need to put the right toString. We can change the reference type to HashMap for clarity if needed.

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ProducerManager.java:
```diff
@@ -0,0 +1,125 @@
+            Exception ex = callback.exception();
+            if (ex instanceof KafkaException) {
+                throw (KafkaException) ex;
+            } else {
```
For readability, it'll be useful to place the positive case under `if` and negative case under `else`, such as:

```
if (callback.exception() == null) {
 ...
} else {
 ...
}
```

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ProducerManager.java:
```diff
@@ -0,0 +1,125 @@
+    private final RemoteLogMetadataTopicPartitioner topicPartitioner;
+    private final TopicBasedRemoteLogMetadataManagerConfig rlmmConfig;
+
+    private volatile boolean close = false;
```
Could you pls document the state does this boolean represents?

satishd replied:
Done.

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -0,0 +1,229 @@
+    private final RemotePartitionMetadataEventHandler remotePartitionMetadataEventHandler;
+    private final RemoteLogMetadataTopicPartitioner topicPartitioner;
+
+    private volatile boolean close = false;
```
Could you pls mention what state does this boolean represent?

satishd replied:
Done

ccding replied:
We may want a better variable name here. e.g., `isClosing` or `closed` or something else.

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -0,0 +1,229 @@
+                log.info("Polling consumer to receive remote log metadata topic records");
+                ConsumerRecords<byte[], byte[]> consumerRecords
+                        = consumer.poll(Duration.ofSeconds(POLL_INTERVAL_MS));
+                for (ConsumerRecord<byte[], byte[]> record : consumerRecords) {
```
Is it useful to assert that `record.key()` is empty before the key is ignored below?

satishd replied:
I do not think that check is really needed here.

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -0,0 +1,229 @@
+            log.error("Error occurred in consumer task, close:[{}]", close, e);
+        }
+
+        closeConsumer();
```
Hmm, should you be setting `close` to true here? (it depends on the meaning of `close`, which I don't fully understand....)

satishd replied:
It indicates whether the closing process has been started or not. If it is set as true, consumer will stop consuming messages and it will not allow partition assignments to be updated.
Updated the java doc of close.

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -0,0 +1,229 @@
+    private volatile boolean close = false;
+    private volatile boolean assignPartitions = false;
+
+    private final Object assignPartitionsLock = new Object();
```
Do you really need this explicit lock? It seems you could just use `wait()` and `notify()` APIs on the `ConsumerTask` object instead, combined with `synchronized` keyword.

satishd replied:
I wanted to have a separate lock instance specifically for the assignments and the respective processing. It gives better clarity and separations even if we add any other logic by taking lock on this instance.

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataTopicPartitioner.java:
```diff
@@ -0,0 +1,60 @@
+    public static final Logger log = LoggerFactory.getLogger(RemoteLogMetadataTopicPartitioner.class);
+    private final int noOfMetadataTopicPartitions;
+
+    public RemoteLogMetadataTopicPartitioner(int noOfMetadataTopicPartitions) {
```
s/noOfMetadataTopicPartitions/numMetadataTopicPartitions

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataTopicPartitioner.java:
```diff
@@ -0,0 +1,60 @@
+    public int metadataPartition(TopicIdPartition topicIdPartition) {
+        Objects.requireNonNull(topicIdPartition, "TopicPartition can not be null");
+
+        int partitionNo = Utils.toPositive(Utils.murmur2(toBytes(topicIdPartition))) % noOfMetadataTopicPartitions;
```
s/partitionNo/partition

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataTopicPartitioner.java:
```diff
@@ -0,0 +1,60 @@
+        Objects.requireNonNull(topicIdPartition, "TopicPartition can not be null");
+
+        int partitionNo = Utils.toPositive(Utils.murmur2(toBytes(topicIdPartition))) % noOfMetadataTopicPartitions;
+        log.debug("No of partitions [{}], partitionNo: [{}] for given topic: [{}]", noOfMetadataTopicPartitions, partitionNo, topicIdPartition);
```
s/No of.../Num of...

Also it feels overkill to me to log this message for each call.

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataTopicPartitioner.java:
```diff
@@ -0,0 +1,60 @@
+        // We do not want to depend upon hash code generation of Uuid as that may change.
+        int hash = Objects.hash(topicIdPartition.topicId().getLeastSignificantBits(),
+                                topicIdPartition.topicId().getMostSignificantBits(),
+                                topicIdPartition.topicPartition().topic(),
```
Is topic necessary here, when UUID and partition is already sufficient input for the hash?

satishd replied:
Good point. We can skip topic as topic-id is sufficient here.

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemotePartitionMetadataStore.java:
```diff
@@ -0,0 +1,140 @@
+public class RemotePartitionMetadataStore extends RemotePartitionMetadataEventHandler implements Closeable {
+    private static final Logger log = LoggerFactory.getLogger(RemotePartitionMetadataStore.class);
+
+    private Map<TopicIdPartition, RemotePartitionDeleteMetadata> idToPartitionDeleteMetadata =
```
It seems these 2 attributes can be marked final if you call `Map.clear()` in `close()` instead of replacing the reference.

satishd replied:
I went with assigning empty map as Map.clear() needs to go through all the entries and dereference them. Another way is to leave the map as it is and set the cose state and do not allow any operation when it is closed but it will have a check for each call.

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemotePartitionMetadataStore.java:
```diff
@@ -0,0 +1,140 @@
+            try {
+                remoteLogMetadataCache.updateRemoteLogSegmentMetadata(rlsmUpdate);
+            } catch (RemoteResourceNotFoundException e) {
+                log.error("Error occurred while updating the remote logs segment.");
```
s/logs/log ?

kowshik commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManager.java:
```diff
@@ -0,0 +1,335 @@
+
+    // Using AtomicBoolean instead of volatile as it may encounter http://findbugs.sourceforge.net/bugDescriptions.html#SP_SPIN_ON_FIELD
+    // if the field is read but not updated in a spin loop like in #initializeResources() method.
+    private final AtomicBoolean close = new AtomicBoolean(false);
```
Could you pls add a comment on what state does `close` represent?

satishd replied:
Done

kowshik reviewed (commented):
@satishd Thanks for the PR. I've made a pass on non-test code. Some comments below.

satishd committed:
Addressed review comments.

satishd commented:
Thanks @junrao @kowshik for the review comments. Addressed them with the latest commits.

ccding commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java:
```diff
@@ -0,0 +1,139 @@
+        final long offset = recordMetadata.offset();
+        long startTimeMillis = time.milliseconds();
+        while (true) {
+            long committedOffset = committedOffset(partition);
```
Can we avoid making the variable name the same as the function name?

ccding commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java:
```diff
@@ -0,0 +1,139 @@
+        consumerTask.addAssignmentsForPartitions(allPartitions);
+    }
+
+    public void removeAssignmentsForPartitions(Set<TopicIdPartition> partitions) {
```
why is the variable name `partitions`, while the one above for `addAssignmentsForPartitions` is `allPartitions`?

also why one is `Set` and the other is `HashSet`?

ccding commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java:
```diff
@@ -0,0 +1,139 @@
+        }
+
+        final long offset = recordMetadata.offset();
+        long startTimeMillis = time.milliseconds();
```
I think in the codebase we use `Ms` more often than using `Millis`

ccding commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java:
```diff
@@ -0,0 +1,139 @@
+public class ConsumerManager implements Closeable {
+
+    private static final Logger log = LoggerFactory.getLogger(ConsumerManager.class);
+    private static final long CONSUME_RECHECK_INTERVAL_MS = 50L;
```
should this be fixed or configurable in `rlmmConfig`?

satishd replied:
I plan to have a config with a default value in a follow-up PR.

ccding commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -0,0 +1,231 @@
+    private final RemotePartitionMetadataEventHandler remotePartitionMetadataEventHandler;
+    private final RemoteLogMetadataTopicPartitioner topicPartitioner;
+
+    // It indicates whether the closing process has been started or not. Id it is set as true,
```
`Id` -> `If`

ccding commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -0,0 +1,231 @@
+    public void run() {
+        log.info("Started Consumer task thread.");
+        try {
+            while (!close) {
```
multiple threads are reading and writing `close`, which is not thread safe

satishd replied:
`close/closing` is already volatile and its state is immediately reflected in other threads. 
In `close()` method, the consumer is invoked with `wakeup()`, and the other thread may receive `WakeupException` if it is executing `poll()` or earlier than that. If it is after `poll` then the next check of `close/closing` allows to come out and finally close the consumer.

`updateAssignmentsForPartitions` and `close()` methods can not be called concurrently as  it is alreadyhandled by `TopicBasedRemoteLogMetadataManager`.

ccding commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -0,0 +1,231 @@
+class ConsumerTask implements Runnable, Closeable {
+    private static final Logger log = LoggerFactory.getLogger(ConsumerTask.class);
+
+    private static final long POLL_INTERVAL_MS = 30L;
```
why is this set to 30, rather than another number like 10, 50, 100?

satishd replied:
This is a temporary change. I plan to have a config with a default value.

ccding commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -0,0 +1,231 @@
+
+    private void closeConsumer() {
+        log.info("Closing the consumer instance");
+        if (consumer != null) {
```
is the check necessary?

ccding commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -0,0 +1,231 @@
+    // It indicates whether the closing process has been started or not. Id it is set as true,
+    // consumer will stop consuming messages and it will not allow partition assignments to be updated.
+    private volatile boolean close = false;
+    private volatile boolean assignPartitions = false;
```
can you explain what this variable means?

ccding commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -0,0 +1,231 @@
+        updateAssignmentsForPartitions(updatedPartitions, Collections.emptySet());
+    }
+
+    public void removeAssignmentsForPartitions(Set<TopicIdPartition> partitions) {
```
also here. Why are the variable names different? One is `updatedPartitions` and the other is `partitions`

satishd replied:
When the code was refactored, it went with the caller method arg names. thanks for catching these.

ccding commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -0,0 +1,231 @@
+        return Optional.ofNullable(partitionToConsumedOffsets.get(partition));
+    }
+
+    public boolean assignedPartition(int partition) {
```
The function name is a little confusing with variable names in the same class, e.g., `assignPartitions`, `assignedTopicPartitions`. I am not sure if it would be better to rename `assignedPartition` to `isAssignedPartition`.

satishd replied:
What about `isPartitionAssigned`?

ccding commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -0,0 +1,231 @@
+    }
+
+    public void close() {
+        if (!close) {
```
not good to use the same name for a variable and a function.

ccding commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ProducerManager.java:
```diff
@@ -0,0 +1,132 @@
+            if (metadataPartitionNum >= rlmmConfig.metadataTopicPartitionsCount()) {
+                // This should never occur as long as metadata partitions always remain the same.
+                throw new KafkaException("Chosen partition no " + metadataPartitionNum +
+                                         " is more than the partition count: " + rlmmConfig.metadataTopicPartitionsCount());
```
greater than or equal to?

satishd replied:
What about `must be less than the partition count`, conveys the intent clearly.

ccding commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ProducerManager.java:
```diff
@@ -0,0 +1,132 @@
+        try {
+            if (metadataPartitionNum >= rlmmConfig.metadataTopicPartitionsCount()) {
+                // This should never occur as long as metadata partitions always remain the same.
+                throw new KafkaException("Chosen partition no " + metadataPartitionNum +
```
will the exception be caught by your own catch in line 74? maybe move it out of the `try` block?

ccding commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ProducerManager.java:
```diff
@@ -0,0 +1,132 @@
+            return callback.recordMetadata();
+        } else {
+            Exception ex = callback.exception();
+            if (ex instanceof KafkaException) {
```
out of curiosity: why we don't do the check within the `new KafkaException(...)` call

satishd replied:
If you are asking about the earlier catch block, that will not cover the exceptions like receiving from broker etc.  The earlier catch block is applicable only until the record is added to the accumulator.

ccding commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ProducerManager.java:
```diff
@@ -0,0 +1,132 @@
+
+        close = true;
+
+        if (producer != null) {
```
is the check necessary?

ccding commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogMetadataTopicPartitioner.java:
```diff
@@ -0,0 +1,59 @@
+            (byte) n
+        };
+    }
+}
```
new line at the end of the file

ccding commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManager.java:
```diff
@@ -0,0 +1,336 @@
+    }
+
+    @Override
+    public void putRemotePartitionDeleteMetadata(RemotePartitionDeleteMetadata remotePartitionDeleteMetadata)
```
function names `addRemoteLogSegmentMetadata`, `updateRemoteLogSegmentMetadata`, and `putRemotePartitionDeleteMetadata` don't seem very consistent. Is there a way to improve it?

satishd replied:
Javadocs explain the behaviorr in detail.
`addRemoteLogSegmentMetadata` - adds a new entry. 
`updateRemoteLogSegmentMetadata ` - updates an existing entry. 
`putRemotePartitionDeleteMetadata ` - adds or updates an existing entry, put is generally used for that purpose. If this is not so clear, another option can be `addOrUpdateRemotePartitionDeleteMetadata`.

ccding commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManager.java:
```diff
@@ -0,0 +1,336 @@
+        Objects.requireNonNull(topicIdPartition, "topicIdPartition can not be null");
+        ensureInitializedAndNotClosed();
+
+        return remotePartitionMetadataStore.listRemoteLogSegments(topicIdPartition, leaderEpoch);
```
Here has a race condition. It is possible `close=false` before calling `ensureInitializedAndNotClosed()` and the `close()` function call by another thread has completed before calling `remotePartitionMetadataStore.listRemoteLogSegments(topicIdPartition, leaderEpoch)`.

I think we should always grab a read lock before calling `ensureInitializedAndNotClosed();`, or do some other fancy things.

satishd replied:
I did not add the checks for these methods as they will not be invoked in parallel when close is called. But I agree to have the guards here with read lock.

satishd committed:
Addressed review comments
  - Added a few method/variable names and improved locking.

satishd commented:
Thanks @ccding for the review. Addressed them with the latest commit https://github.com/apache/kafka/commit/25c4ccb7c5ad44963a9708f0df67b982cf91ac15

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java:
```diff
@@ -0,0 +1,130 @@
+        //Create a task to consume messages and submit the respective events to RemotePartitionMetadataEventHandler.
+        KafkaConsumer<byte[], byte[]> consumer = new KafkaConsumer<>(rlmmConfig.consumerProperties());
+        consumerTask = new ConsumerTask(consumer, remotePartitionMetadataEventHandler, rlmmTopicPartitioner);
+        consumerTaskThread = KafkaThread.daemon("RLMMConsumerTask", consumerTask);
```
Hmm, why is this a daemon thread? It seems that we want to coordinate the shutdown of the thread.

satishd replied:
Good catch, updated it.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java:
```diff
@@ -0,0 +1,130 @@
+        if (!consumerTask.isPartitionAssigned(partition)) {
+            log.warn("This consumer is not subscribed to the target partition [{}] on which message is produced.",
+                    partition);
+            return;
```
Hmm, should we throw an exception in this case so that the caller knows the operation has failed?

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java:
```diff
@@ -0,0 +1,130 @@
+                    offset, partition, committedOffset, CONSUME_RECHECK_INTERVAL_MS);
+
+            if (time.milliseconds() - startTimeMs > rlmmConfig.consumeWaitMs()) {
+                log.warn("Committed offset for partition:[{}] is : [{}], but the target offset: [{}] ",
```
Similar to the above,  should we throw an exception in this case so that the caller knows the operation has failed?

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -0,0 +1,235 @@
+                // If no partitions are assigned, wait until they are assigned.
+                log.info("Waiting for assigned remote log metadata partitions..");
+                try {
+                    assignPartitionsLock.wait();
```
We need to unblock the wait if we are closing the consumer.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -0,0 +1,235 @@
+    private void updateAssignmentsForPartitions(Set<TopicIdPartition> addedPartitions,
+                                                Set<TopicIdPartition> removedPartitions) {
+        log.info("Updating assignments for addedPartitions: {} and removedPartition: {}", addedPartitions, removedPartitions);
+        ensureNotClosed();
```
Is this necessary since immediately after this check, the consumer task could be closed. Ditto in other places.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ProducerManager.java:
```diff
@@ -0,0 +1,130 @@
+    }
+
+    public RecordMetadata publishMessage(RemoteLogMetadata remoteLogMetadata) throws KafkaException {
+        ensureNotClosed();
```
Is this necessary? Once producer is closed, send() will throw an exception.

junrao commented on storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogSegmentLifecycleManager.java:
```diff
@@ -0,0 +1,60 @@
+ * different implementations of this interface. This is responsible for managing all the segments for a given {@code topicIdPartition}
+ * registered with {@link #initialize(TopicIdPartition)}.
+ */
+public interface RemoteLogSegmentLifecycleManager extends Closeable {
```
How is this different from RemoteLogMetadataCache? It seems that it's just a wrapper over RemoteLogMetadataCache?

satishd replied:
There are two implementations about this class for both `RemoteLogMetadataCache` and `TopicBasedRemoteLogMetadataManager`. There are common tests in `RemoteLogSegmentLifecycleTest `that we want to run for both of them.

junrao replied:
Is RemoteLogSegmentLifecycleManager used for tests only? If so, could we move it under tests?

satishd replied:
`RemoteLogSegmentLifecycleManager` is already under tests.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemotePartitionMetadataStore.java:
```diff
@@ -0,0 +1,140 @@
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class RemotePartitionMetadataStore extends RemotePartitionMetadataEventHandler implements Closeable {
```
Could we add a comment for this class?

satishd replied:
Updated with a comment.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManager.java:
```diff
@@ -0,0 +1,360 @@
+            // Scheduling the initialization producer/consumer managers in a separate thread. Required resources may
+            // not yet be available now. This thread makes sure that it is retried at regular intervals until it is
+            // successful.
+            initializationThread = KafkaThread.daemon("RLMMInitializationThread", () -> initializeResources());
```
Why do we need to initialize in a separate thread?

satishd replied:
It was required to be retried until the topic is successfully created. I added the logic to check for topic creation too.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManager.java:
```diff
@@ -0,0 +1,360 @@
+    // It indicates whether the close process of this instance is started or not via #close() method.
+    // Using AtomicBoolean instead of volatile as it may encounter http://findbugs.sourceforge.net/bugDescriptions.html#SP_SPIN_ON_FIELD
+    // if the field is read but not updated in a spin loop like in #initializeResources() method.
+    private final AtomicBoolean closing = new AtomicBoolean(false);
```
Since we have a lock, could we just make closing a boolean?

satishd replied:
`closing` is accessed in `initializeResources` and we do not need to take a lock there.  I would like to keep this as `AtomicBoolean` which addresses that and it is easy to understand the semantics.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManagerConfig.java:
```diff
@@ -0,0 +1,200 @@
+
+    public static final String REMOTE_LOG_METADATA_TOPIC_REPLICATION_FACTOR_PROP = "remote.log.metadata.topic.replication.factor";
+    public static final String REMOTE_LOG_METADATA_TOPIC_PARTITIONS_PROP = "remote.log.metadata.topic.num.partitions";
+    public static final String REMOTE_LOG_METADATA_TOPIC_RETENTION_MILLIS_PROP = "remote.log.metadata.topic.retention.ms";
```
MILLIS => MS to be consistent with other places. Ditto in a few other places.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManagerConfig.java:
```diff
@@ -0,0 +1,200 @@
+    private Map<String, Object> producerProps;
+
+    public TopicBasedRemoteLogMetadataManagerConfig(Map<String, ?> props) {
+        log.info("Received props: [{}]", props);
```
Is this needed since we log all configs when creating KafkaConfig already?

junrao commented on storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManagerHarness.java:
```diff
@@ -0,0 +1,123 @@
+        return topicBasedRemoteLogMetadataManager;
+    }
+
+    private void createMetadataTopic() {
```
It seems that we need to automatically create metadata topic in RLMM implementation, not just in tests.

satishd replied:
I was planning to add that later as mentioned earlier. I updated with the required changes in latest commit.

junrao replied:
createMetadataTopic() is no longer used.

junrao commented on storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManagerHarness.java:
```diff
@@ -0,0 +1,123 @@
+
+        topicBasedRemoteLogMetadataManager.configure(configs);
+        try {
+            waitUntilInitialized(120_000);
```
120s seems quite long. Do we need to wait that long?

satishd replied:
I guess 60s may be sufficient, updated with that.

junrao commented on storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogSegmentLifecycleTest.java:
```diff
@@ -0,0 +1,516 @@
+            Assertions.assertFalse(remoteLogSegmentLifecycleManager.remoteLogSegmentMetadata(40, 1).isPresent());
+
+            // Check that these leader epochs are to considered for highest offsets as they are still getting copied and
+            // they did nto reach COPY_SEGMENT_FINISHED state.
```
typo nto

junrao commented on storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/RemoteLogSegmentLifecycleTest.java:
```diff
@@ -0,0 +1,516 @@
+            // it reaches RemoteLogSegmentState.COPY_SEGMENT_FINISHED.
+            Assertions.assertFalse(remoteLogSegmentLifecycleManager.remoteLogSegmentMetadata(40, 1).isPresent());
+
+            // Check that these leader epochs are to considered for highest offsets as they are still getting copied and
```
This sentence doesn't read well.

junrao commented on storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManagerConfigTest.java:
```diff
@@ -0,0 +1,149 @@
+            Assertions.assertEquals(entry.getValue(),
+                                    rlmmConfig.producerProperties().get(entry.getKey()));
+            Assertions.assertEquals(entry.getValue(),
+                                    rlmmConfig.producerProperties().get(entry.getKey()));
```
This should be for consumer?

junrao commented on storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManagerTest.java:
```diff
@@ -0,0 +1,132 @@
+                    break;
+                }
+            } else {
+                if (topicBasedRlmm().receivedOffsetForPartition(leaderMetadataPartition).orElse(-1L) > -1 ||
```
Why are we testing against -1 here but 0 above?

satishd replied:
Updated with a comment in the code.

junrao reviewed (commented):
@satishd : Thanks for the updated PR. A few more comments.

satishd committed:
- Addressed review comments.
 - Introduced exceptions instead of ignoring in the cases like
    - timeouts
    - if the published message's partition is not yet subscribed by RLMM yet
 - Handled race conditions of initializing RLMM and handling leadership changes and stop-partition callbacks.
 - Minor cleanups.

satishd commented:
Thanks @junrao for your comments. Addressed them with the latest commit.

satishd commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManager.java:
```diff
@@ -0,0 +1,402 @@
+    }
+
+    @Override
+    public void onPartitionLeadershipChanges(Set<TopicIdPartition> leaderPartitions,
```
There can be a race here between this method and initializeResources() which I plan to address in a followup PR. I am working on a separate consumer to act on assigned partitions to cathcup. This race is being handled with those changes., same applies to onStopPartitions method too.

ccding commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -220,16 +227,16 @@ public boolean isPartitionAssigned(int partition) {
+                // Closing should be updated only after acquiring the lock to avoid race in
+                // maybeWaitForPartitionsAssignment() where it waits on assignPartitionsLock. It should not wait
+                // if the closing is already set.
+                closing = true;
```
Is it possible two threads call `close()` concurrently?

satishd replied:
`close()` will not be called concurrently here.

ccding commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManager.java:
```diff
@@ -51,6 +58,7 @@
  */
 public class TopicBasedRemoteLogMetadataManager implements RemoteLogMetadataManager {
     private static final Logger log = LoggerFactory.getLogger(TopicBasedRemoteLogMetadataManager.class);
+    private static final long INITIALIZATION_RETRY_INTERVAL_MS = 30000L;
```
our of curiosity: why we have `L` here but not https://github.com/apache/kafka/pull/10579/commits/77da4ebff5228d1c8938105d50b6b15a5223b294#diff-9f35ec72323f704125b5ee6d35a230de7f28f8c448a706eec2946fa46a12cfedR63?

satishd replied:
I prefer adding L for longs, which I missed at other declaration. afaik, that does not cause any issues as it gets automatically converted via a widening conversion to a `long`. The compiler takes care of not allowing numbers that may get truncated from `int` to `long` widening. Thanks for catching it, I will make it consistent by adding it.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:
```diff
@@ -0,0 +1,242 @@
+                return;
+            }
+
+            while (assignedMetaPartitions.isEmpty() && !closing) {
```
The test for closing seem unnecessary since closing can't change while synchronized on assignPartitionsLock.

satishd replied:
This loop can run after `assignPartitionsLock.wait()` call which might have been notified from `close()` method, we should have a `closing` check to get out of the loop. 

We can have a more aggressive check to return from here when `closing` is true, I will add that.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManager.java:
```diff
@@ -0,0 +1,432 @@
+            // Scheduling the initialization producer/consumer managers in a separate thread. Required resources may
+            // not yet be available now. This thread makes sure that it is retried at regular intervals until it is
+            // successful.
+            initializationThread = KafkaThread.nonDaemon("RLMMInitializationThread", () -> initializeResources());
```
> It was required to be retried until the topic is successfully created. I added the logic to check for topic creation too.

I am still not sure why we need to initialize in a separate thread. If we can't create the metadata topic or instantiate the producer/consumer due to wrong configurations, we want to fail fast by throwing an error to shut down the broker.

satishd replied:
@junrao The reason why we need to initialize in a different thread here is that RLMM will be created and `configure()` will be called before the broker starts accepting the requests. So, we can not call topic creation requests in the same thread as the brokers are not yet up. 

Another way to do this is to have this topic as an internal topic and it will be auto created whenever it is accessed. afaik, creating producer and consumer instances can be done without the brokers up and running and they will not trigger a request to auto creation of remote log metadata topic. Consumer assignment will send a metadata request which will trigger auto creation of topics. This assignment on consumer is done only when RLMM receives callback thorough the broker about leader and isr updates from the controller. This is what we had in pre-2.7 implementation but we saw an intermittent deadlock issue but I do not see that happening on trunk.  

In the current PR, I will make the existing producer manager and consumer manager in configure() and have the topic creation done in the tests. 

I will have a quick follow-up PR with internal topic changes and remove the topic creation code from tests.

satishd replied:
I made the mentioned changes for this PR in the latest commit.

junrao replied:
@satishd : To create a topic, you don't need a particular broker to be up. You just need to be able to access the bootstrap brokers. Auto creating this topic on access is a bit weird since it's not truly an internal topic and it is just an implementation detail of RLMM. So, it makes more sense for topic-based RLMM to create it.

satishd replied:
@junrao 
As I mentioned earlier, RLMM is created before broker starts accepting the requests. So, when RLMM is getting initialized and tries to create a topic in the same thread, then none of the brokers(including the brokers related to the bootstrap-servers config) will be available for taking any admin client requests for topic creation. That is why I was doing this in a different thread as it allows the broker/controller to comeup and accept the admin client request for topic creation.

junrao replied:
@satishd : My understanding is that when we enable remote storage, we will do that through a rolling upgrade. So, at any given time, there is at most a single broker being down. Therefore, as long as the bootstrap broker list contains more than one broker, operations like creating topics can still be done while a single broker is starting.

satishd replied:
@junrao: That is a good point. But that is valid only for upgrade scenario.

There are two scenarios here.
1) Upgrade path
2) Fresh install/deploy with the release containing this feature.

What you said makes sense for upgrade path. But "bootstrap.servers" list is configured with the local broker endpoint only in which RLMM is getting initialized. So, if we try to initialize RLMM in the same thread, it wont be able to connect to the local broker as it is not yet started to accept the broker API requests. One way to address this is to give `bootstrap.servers` with more than one broker's endpoint. This will also put a limitation to create a cluster with one broker instance for test/demo environments. 

When users install/deploy a fresh environment with the release containing this feature, there should not be a restriction to do rolling restarts with the configuration enabled. Please let me know if I am missing anything here.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManager.java:
```diff
@@ -0,0 +1,432 @@
+        } catch (Exception e) {
+            if (e.getCause() instanceof TopicExistsException) {
+                log.info("Topic [{}] already exists", topic.name());
+                topicCreated = true;
```
Should we verify that the number of partitions in the existing topic matches the configuration?

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManager.java:
```diff
@@ -0,0 +1,432 @@
+            if (!createTopic(remoteLogMetadataTopicRequest)) {
+                // try to create the topic again if it could not be created.
+                log.info("Sleep for : {} ms before it is retried again.", INITIALIZATION_RETRY_INTERVAL_MS);
+                Utils.sleep(INITIALIZATION_RETRY_INTERVAL_MS);
```
Hmm, we don't want to retry forever. If there is a configuration error, we want to fail fast.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManager.java:
```diff
@@ -0,0 +1,432 @@
+                // when it is initialized successfully in initializeResources().
+                this.pendingAssignPartitions.addAll(allPartitions);
+            } else {
+                this.pendingAssignPartitions.clear();
```
Hmm, why is this needed since initializeResources() does this already?

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManager.java:
```diff
@@ -0,0 +1,432 @@
+        boolean topicCreated = false;
+        AdminClient adminClient = null;
+        try {
+            adminClient = AdminClient.create(rlmmConfig.consumerProperties());
```
It seems producer credentials are closer for the admin client.

junrao reviewed (commented):
@satishd : Thanks for the updated PR. A few more comments.

satishd committed:
- Addressed review comments.
 - Added configs for initialization retry interval and retry max duration.
 - It throws FatalError if the retries are exhausted with in retry max duration config.

satishd commented:
Thanks @junrao for the comments. Pl take a look at my inline replies and the latest commit.

kamalcph commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ProducerManager.java:
```diff
@@ -0,0 +1,120 @@
+
+    public void close() {
+        // If it is already closed, return from here.
+        if (!closing.compareAndSet(false, true)) {
```
This condition should be inverted.

satishd replied:
Good catch, we do not really need this check here as it is already guarded in `TopicBasedRemoteLogMetadataManager`.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManager.java:
```diff
@@ -0,0 +1,484 @@
+                if (time.milliseconds() - startTimeMs > rlmmConfig.initializationRetryMaxTimeoutMs()) {
+                    log.error("Timed out in initializing the resources, retried to initialize the resource for [{}] ms.",
+                              rlmmConfig.initializationRetryMaxTimeoutMs());
+                    initializationFailed = true;
```
Should halt the jvm in this case? Ditto below when the partition count doesn't match.

satishd replied:
Yes, we are already doing that [here](https://github.com/apache/kafka/pull/10579/files#diff-50bf08ff4e92c16dfbea9239c89467a3ebf5fa38e5f4ea44ed4acff80013136eR452) whenever this class is accessed with/for remote log metadata operations.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManager.java:
```diff
@@ -0,0 +1,484 @@
+    private final RemotePartitionMetadataStore remotePartitionMetadataStore = new RemotePartitionMetadataStore();
+    private volatile TopicBasedRemoteLogMetadataManagerConfig rlmmConfig;
+    private volatile RemoteLogMetadataTopicPartitioner rlmmTopicPartitioner;
+    private final Set<TopicIdPartition> pendingAssignPartitions = Collections.synchronizedSet(new HashSet<>());
```
Does pendingAssignPartitions need to be synchronizedSet since it's accessed under a lock?

satishd replied:
It is needed because `pendingAssignPartitions` is updated in both `onPartitionLeadershipChanges` and `onStopPartitions` methods which can happen concurrently as both of them take read lock.

junrao reviewed (commented):
@satishd : Thanks for the updated PR. Just a few more comments.

satishd commented:
Thanks @junrao for your latest comments, replied to them inline.

junrao commented on storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManager.java:
```diff
@@ -0,0 +1,484 @@
+            }
+
+            // Publish the message to the topic.
+            doPublishMetadata(segmentMetadataUpdate.remoteLogSegmentId().topicIdPartition(), segmentMetadataUpdate);
```
Since producerManager is initialized asynchronously, how do we deal with the case when the producerManager is not ready when an event needs to be published?

satishd replied:
That is a good point. We plan to improve the semantics here. 

Earlier, we plan to introduce RetriableException for RLMM and RSM so that callers can have an option to know whether they can retry or not. In the case of initialization is not yet complete, RetriableException can be thrown caller can retry based on backoff. RLMM can send non retriable exception if it is in closing state and there will not be any retries.  

Another way to handle this is to take these events and store them in in-memory queue and return Future. These Futures will be considered successful if initialization is successful and the events are published to the topic.

I plan to address these in a followup PR while these APIs are integrated with RLM, filed https://issues.apache.org/jira/browse/KAFKA-13097.

junrao reviewed (commented):
@satishd : Thanks for the updated PR. A couple of more comments.

junrao reviewed (commented):
@satishd : Thanks for the reply. One more comment below.

satishd committed:
Addressed review comments
  - Removed "bootstrap.server" config from RLMM and it can be passed with topic based RLMM properties with "remote.log.metadata.common.client." or "remote.log.metadata.producer." or "remote.log.metadata.consumer."

satishd commented:
Thanks @junrao for the comment. I addressed it with the latest commit.

junrao reviewed (approved):
@satishd : Thanks for the updated PR. LGTM
